File: aoc2023/python-solutions/src/day02.py
from collections import defaultdict
from functools import reduce
from util import INPUTS
import re

type game = list[tuple[int, str]]


def parse() -> list[game]:
    with open(f"{INPUTS}/day02.txt", "r") as inp:
        return [
            [(int(m[0]), m[1]) for m in re.findall(r"([0-9]+) (green|blue|red)", line)]
            for line in inp.readlines()
        ]


def game_possible(val: int, color: str) -> bool:
    match color:
        case "blue":
            return 14 >= val
        case "green":
            return 13 >= val
        case "red":
            return 12 >= val
        case _:
            return False


def possible_games(xs: list[game]) -> int:
    return sum(
        [
            i
            for (i, xs) in zip(range(1, len(xs) + 1), xs)
            if all(game_possible(v, c) for (v, c) in xs)
        ]
    )


def part2(cubes: game) -> int:
    res = defaultdict(int)  # Defaults to 0
    for val, color in cubes:
        res[color] = max(res[color], val)
    return reduce(lambda a, b: a * b, res.values())


def day02() -> tuple[int, int]:
    input = parse()
    return possible_games(input), sum([part2(cubes) for cubes in input])


assert day02() == (2776, 68638)

if __name__ == "__main__":
    print(day02())

File: aoc2023/ocaml_solutions/lib/day02.ml
open Core
open Util
open Util.Parse

let possible (n, t) =
  n <= match t with
       | "red" -> 12
       | "green" -> 13
       | "blue" -> 14
       | _ -> -1

let parse =
  let mk_cols = List.map ~f:(fun ms -> (int_of_string ms.(1), ms.(2)))
  in In_channel.read_lines "../../inputs/day02.txt"
     |> List.map ~f:(re_all "(\\d+) (red|green|blue)")
     |> List.map ~f:mk_cols

let part1 =
  parse
  |> List.filter_mapi
       ~f:(fun n g -> if List.for_all ~f:possible g
                      then Some (n + 1)
                      else None)
  |> sum

let part2 =
  let max_in_group group =
    List.map group ~f:fst
    |> List.fold ~init:0 ~f:max
  in
  parse
  |> List.map
       ~f:(fun l -> List.sort_and_group l
                      ~compare:(fun (_, a) (_, b) -> compare_string a b)
                    |> List.map ~f:max_in_group
                    |> List.fold ~init:1 ~f:( * ))
  |> sum

let day02 = string_of_int part1, string_of_int part2
